### PR TITLE
Improve docs and specs for {(Naive)DateTime, Date, Time}.convert/2

### LINKED ISSUE #5919: Document the conversion functions in calendar
/cc @Qqwy

### PR DESCRIPTION
**Pull request by Qqwy**
This PR builds on top of the now-merged #5841. Fixes #5919.

It adds/improves documentation and specs to NaiveDateTime.convert, Date.convert, Time.convert and DateTime.convert.

### CODE REVIEW

**Comment by Qqwy** [comment]
Two problems I am facing while writing the documentation:

- For consistency and being future-proof, the normal `convert` functions return an `{:ok, result}` tuple. `DateTime.convert` and `Time.convert` have no way right now to fail, as conversion is always possible.
Should I therefore remove the `{:error, any}` part from the spec and the documentation? Or not?

- Showing examples of conversions does not really make a lot of sense when converting to the same calendar we're coming from. Does it still make sense to show these no-op examples (see for instance the docs of `DateTime.convert/2` for an example), or rather not?

**Comment by josevalim** [comment]
@Qqwy Yes, let's return tuples. We can leave the possible `:error` type documented.

File: lib/elixir/lib/calendar.ex [hunk 1]
```diff
@@ -1002,8 +1002,10 @@ defmodule Time do
 
   @doc """
   Converts the Time struct to a different calendar.
+  Returns `{:ok, %Time{}}` if the conversion was successfull,
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Please add a new line to start a new paragraph.

File: lib/elixir/lib/calendar.ex [hunk 2]
```diff
@@ -1641,6 +1647,14 @@ defmodule NaiveDateTime do
     end
   end
 
+
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: ✂️

File: lib/elixir/lib/calendar.ex [hunk 3]
```diff
@@ -1641,6 +1647,14 @@ defmodule NaiveDateTime do
     end
   end
 
+  
+  @doc """
+  Converts a NaiveDateTime from one calendar to another.
+  If it is not possible to convert unambiguously between the calendars
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Please add a new line to start a new paragraph.

File: lib/elixir/lib/calendar.ex [hunk 4]
```diff
@@ -1656,6 +1670,12 @@ defmodule NaiveDateTime do
     end
   end
 
+  @doc """
+
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: ✂️

File: lib/elixir/lib/calendar.ex [hunk 5]
```diff
@@ -1656,6 +1670,12 @@ defmodule NaiveDateTime do
...
+  Converts a NaiveDateTime from one calendar to another.
+  If it is not possible to convert unambiguously between the calendars
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Please add a new line to start a new paragraph.

File: lib/elixir/lib/calendar.ex [hunk 6]
```diff
@@ -2290,6 +2310,7 @@ defmodule DateTime do
 
   @doc """
   Converts a DateTime from one calendar to another.
+  If this conversion fails for some reason, an `{:error, reason}` tuple is returned.
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Please add a new line to start a new paragraph.

File: lib/elixir/lib/calendar.ex [hunk 7]
```diff
@@ -2314,6 +2336,21 @@ defmodule DateTime do
     {:ok, result_datetime}
   end
 
+  @doc """
+  Converts a DateTime from one calendar to another.
+  If this conversion fails for some reason, an ArgumentError is raised.
+
+  ## Examples
+
+
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: ✂️

File: lib/elixir/lib/calendar.ex [hunk 8]
```diff
@@ -2314,6 +2336,21 @@ defmodule DateTime do
     {:ok, result_datetime}
   end
 
+  @doc """
+  Converts a DateTime from one calendar to another.
+  If this conversion fails for some reason, an ArgumentError is raised.
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Please add a new line to start a new paragraph. The first paragraph must always be a short summary.

File: lib/elixir/lib/calendar.ex (hunk 1 above)
Comment (whatyouhide) [inline lib/elixir/lib/calendar.ex]: "successfull" -> "successful"

File: lib/elixir/lib/calendar.ex [hunk 9]
```diff
@@ -2314,6 +2336,21 @@ defmodule DateTime do
...
+  ## Examples
+
+
+      iex> dt1 = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                 hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                 utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.convert!(dt1, Calendar.ISO)
+      %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
```
Comment (whatyouhide) [inline lib/elixir/lib/calendar.ex]: Please add a newline after the code sample and before the closing `"""` :)

File: lib/elixir/lib/calendar.ex [hunk 10]
```diff
@@ -1002,8 +1002,11 @@ defmodule Time do
 
   @doc """
   Converts the Time struct to a different calendar.
+
+  Returns `{:ok, %Time{}}` if the conversion was successful,
+  or `{:error, reason}` if it was not, for some reason.
   """
-  @spec convert(Time.t, Calendar.calendar) :: Time.t
+  @spec convert(Time.t, Calendar.calendar) :: {:ok, Calendar.time} | {:error, any}
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: `atom` instead of `any`?

Comment (Qqwy) [inline lib/elixir/lib/calendar.ex]: I was not aware that error results are only ever `atom`s. Will change!

File: lib/elixir/lib/calendar.ex (hunk 9 above)
Comment (Qqwy) [inline lib/elixir/lib/calendar.ex]: Done!

File: lib/elixir/lib/calendar.ex (hunk 10 above)
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: They are not but that's the signature you use in other cases, so we should be consistent. We can extend it later if necessary. :)

**Comment by Qqwy** [comment]
Question about the doctests: Right now, only `DateTime.convert/2` and `DateTime.convert!/2` have doctests. They were only added at the earlier PR to show what a doctest for these functions might look like.
It is weird to have them as we cannot convert to another calendar, because only Calendar.ISO is loaded.

Three options:

- Show examples but with moot conversions from `Calendar.ISO` to `Calendar.ISO`.
- Have doctests that include the line `Code.require_file "lib/elixir/test/elixir/fixtures/calendar/julian.exs", __DIR__`, although this might not work when not running Elixir from source (?). Another way to only include this alternative calendar for these doctests?
- Do not have doctests for these functions.

_(Note that, regardless of the doctests, there are normal tests that make sure that the conversions (continue to) work as intended.)_

**Comment by josevalim** [comment]
@Qqwy the goal of doctests is to be able to reproducible by the user, if they cannot be reproducible, it is OK fine, but you need to say so explicitly, such "Imagine an implementation of the Julian calendar".

**Comment by Qqwy** [comment]
@josevalim When someone implements their own Calendar, they will include calls to `(Naive)(Date)(Time).convert/2` in their examples there to show how their calendar could be used. So I guess it is fine to not have doctests here...?

**Comment by josevalim** [comment]
@Qqwy Well, they are implementing the backend, so they likely wouldn't have *those* examples on the doctest.

**Comment by Qqwy** [comment]
> @Qqwy Well, they are implementing the backend, so they likely wouldn't have _those_ examples on the doctest.

I was thinking along the lines of 'If you want to find out what your birthday is in the Chinese calendar, you can call `your_birthday |> Date.convert(Calendar.Chinese)`.

Anyway, still in dubio about this: Should we keep the docs as they are, remove the doctests at DateTime or introduce similar doctests for Time, Date and NaiveDateTime?

**Comment by josevalim** [comment]
@Qqwy I would use Calendar.Julian in the doctests but add a note to the doctests that say: "Imagine someone implements Calendar.Julian".

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**